### PR TITLE
Add basic tidyr/dplyr

### PR DESCRIPTION
**Pull request by KrisanatA**

### CODE REVIEW

**Comment by KrisanatA** [comment]
@mitchelloharawild I would like to add the quiz type in the arrange function section. Is the multiple-choice quiz available?

**Comment by mitchelloharawild** [comment]
Yes, you'll need to cherry pick https://github.com/numbats/monash-learnr-modules/commit/750da0cf07d3d3b0a11c6d84320ac4096067f79e

```
:::{.quiz-multichoice}
- [ ] wrong
- [x] right
- [x] right
:::
```

Is how it works. It visually won't look good yet, nor function perfectly, but the code shouldn't change as I make further improvements to the quiz side of things.

**Comment by KrisanatA** [comment]
@mitchelloharawild Could you check the tip I have written here [64c211b](https://github.com/numbats/monash-learnr-modules/pull/20/commits/64c211b25988aecb90c1a2807ef074b7e51e6c9c). To summarize, I just want to make sure whether .groups = "drop" is the same as using ungroup() function?

File: README.md
Comment (mitchelloharawild) [inline README.md:1]: Unnecessary changes.

File: modules/data-wrangling/index.qmd [hunk 1]
```diff
@@ -69,6 +33,10 @@ Did you know you can also load specific library separately, for example, `librar
 * `count()`
 * `case_when()`
 
+:::{.callout-note}
+This is a very big topic, and we will not cover all of the useful functions in this chapter. Here is a cheat sheet to help you learn more about [dplyr](https://github.com/rstudio/cheatsheets/blob/main/data-transformation.pdf).
```
Comment (mitchelloharawild) [inline modules/data-wrangling/index.qmd:37]: This sounds apologetic - instead I think it could be more focused on what the section covers rather than what it doesn't and emphasise the cheat sheet.

For example:

```md
::: {.callout-note}
## Cheatsheet

This section covers the basic and commonly used dplyr functions for data manipulation. A more comprehensive summary of the functions in the dplyr package can be found in the [dplyr cheatsheet](https://github.com/rstudio/cheatsheets/blob/main/data-transformation.pdf).
:::
```

File: modules/data-wrangling/index.qmd [hunk 2]
```diff
@@ -110,7 +78,10 @@ penguins |>
 
 ???
 
-!all(c("body_mass_g", "year") %in% all.vars(.code))
+c(
+    "Did you use the right function?" != search_ast(.code, .fn = select),
```
Comment (mitchelloharawild) [inline modules/data-wrangling/index.qmd:82]: This isn't working because you are comparing the hint with the result of `search_ast()`.

It should instead be: 

```suggestion
    "Did you use the right function?" = !search_ast(.code, .fn = select),
```

File: modules/data-wrangling/index.qmd [hunk 3]
```diff
@@ -120,31 +91,63 @@ There are many helper functions that allow you to select variables by matching t
...
+#| eval: false
 penguins |>
     select(starts_with("bill"))
 ```
 
+```{r}
+#| echo: false
+penguins |>
+    select(starts_with("bill")) |>
+    head()
```
Comment (mitchelloharawild) [inline modules/data-wrangling/index.qmd:103]: Using `head()` changes the number of rows of the data in the output.
Is the default 10 rows too much? You can change that with `print(data, n=6)` if you must.

You can also set the default behaviour with:
```
options(tibble.print_min = 3L)
```

File: modules/data-wrangling/index.qmd [hunk 4]
```diff
@@ -165,11 +168,301 @@ For the expression field, there are many functions and operators that are useful
...
+#| echo: false
+penguins |>
+    filter(species == "Gentoo") |>
+    head()
+```
+
+
+## arrange()
+
+It is for sorting your data based on one or more variables. By default, it will arrange your data in ascending order. If you want to sort by descending order, it can be done using the `desc()` function.
```
Comment (mitchelloharawild) [inline modules/data-wrangling/index.qmd:186]: Maybe start with an example in asc order, then later show desc order

File: modules/data-wrangling/index.qmd [hunk 5]
```diff
@@ -165,11 +168,301 @@ For the expression field, there are many functions and operators that are useful
...
+
+penguins |>
+    arrange(desc(bill_length_mm)) |>
+    head()
+```
+
+::: {.callout-tip}
+## Did you know
+
+What do you think will happen if you sort the penguin's data by the species where it is a character variable? If you guess that it will sort in **alphabetical order**, you are correct!!!
```
Comment (mitchelloharawild) [inline modules/data-wrangling/index.qmd:211]: More formal writing is preferred.

Comment (mitchelloharawild) [reply modules/data-wrangling/index.qmd:211]: Another common confusion is the sorting of numbers stored as text, for example `c("1", "3", "10", "14")`.
Although this might belong better elsewhere (basics of R objects maybe?)

File: modules/data-wrangling/index.qmd [hunk 6]
```diff
@@ -165,11 +168,301 @@ For the expression field, there are many functions and operators that are useful
...
+```
+
+The **"variable_name"**, this can be the existing variable you are going to modify or  new variable name you are trying to create. The **"value"** is what will be assigned to the **"variable_name"**.
+
+Here is an example on using mutata function to modify the **existing** variable
+
+```{r}
+#| eval: false
+penguins |>
+    mutate(sex = ifelse(sex == "male", "M", "F"))
```
Comment (mitchelloharawild) [inline modules/data-wrangling/index.qmd:233]: `case_when()` or `case_match()` preferred. If you must, `if_else()` is from dplyr and is type safe.

File: modules/data-wrangling/index.qmd [hunk 7]
```diff
@@ -165,11 +168,301 @@ For the expression field, there are many functions and operators that are useful
...
+
+## mutate()
+
+This function is for creating new variable or modify the existing one.
+
+```{r}
+#| eval: false
+
+data |>
+    mutate(variable_name = value)
```
Comment (mitchelloharawild) [inline modules/data-wrangling/index.qmd:223]: More 'real' examples help learning. Could you calculate something else from the penguins dataset?

File: modules/data-wrangling/index.qmd [hunk 8]
```diff
@@ -165,11 +168,301 @@ For the expression field, there are many functions and operators that are useful
...
+
+## group_by()
+
+It allows you to take the existing data and group the specific variables together for future operations. For example, with penguins data, if you want to rank penguins by weights within each species, instead of splitting the dataset by species, rank them individually and then rejoin the data. You can use the `group_by` function together with `mutate` to achieve the same operation.
+
+```{r}
+#| eval: false
+penguins |>
+    group_by(species) |>
+    mutate(ranking = rank(body_mass_g, ties.method = "min")) |>
```
Comment (mitchelloharawild) [inline modules/data-wrangling/index.qmd:277]: An alternative common `group_by() |> mutate()` operation is to calculate percentages with `n/sum(n)`.

Although the emphasis on `mutate()` in the group by section is surprising to me here. I think `group_by()` could be shown later, or alongside `summarise()` with emphasis that the groups also work with other functions.

Something like sections of:

* `summarise()`
* `group_by()` with `summarise()` examples, and a callout about using `group_by()` with other functions (`mutate()`, `filter()`, etc.)

File: modules/data-wrangling/index.qmd [hunk 9]
```diff
@@ -165,11 +168,301 @@ For the expression field, there are many functions and operators that are useful
...
+    ungroup() |>
+    select(species, body_mass_g, ranking) |>
+    head()
+```
+
+:::{.callout-note}
+It is a good practice to use the `ungroup` function after you have done the operations. If you have not **ungrouped** the data, whatever operations you do after will take into account the grouping whether you want it or not.
```
Comment (mitchelloharawild) [inline modules/data-wrangling/index.qmd:291]:
```suggestion
It is a good practice to use the `ungroup()` function after you have done the operations. If you have not **ungrouped** the data, whatever operations you do after will take into account the grouping whether you want it or not.
```

While `ungroup` is technically correct, in my experience new learners need stronger separation of objects and functions. Them being the same is a more advanced topic.

File: modules/data-wrangling/index.qmd [hunk 10]
```diff
@@ -165,11 +168,301 @@ For the expression field, there are many functions and operators that are useful
...
+:::
+
+
+## summarize()
```
Comment (mitchelloharawild) [inline modules/data-wrangling/index.qmd:295]: Australian English is preferred for the Monash modules.
```suggestion
## summarise()
```

File: modules/data-wrangling/index.qmd [hunk 11]
```diff
@@ -165,11 +168,301 @@ For the expression field, there are many functions and operators that are useful
...
+```{r}
+penguins |>
+    group_by(species) |>
+    summarize(
+        Average_body_mass = mean(body_mass_g, na.rm = TRUE),
+        Median_body_mass = median(body_mass_g, na.rm = TRUE),
+        SD_body_mass = sd(body_mass_g, na.rm = TRUE))
+```
+
+:::{.callout-note}
```
Comment (mitchelloharawild) [inline modules/data-wrangling/index.qmd:307]:
```suggestion
:::{.callout-tip}
## Calculations with missing values
```

Callouts should have titles.

File: modules/data-wrangling/index.qmd [hunk 12]
```diff
@@ -165,11 +168,301 @@ For the expression field, there are many functions and operators that are useful
...
+The `na.rm` argument in the function is for ignoring the **NAs** values when calculating each statistic.
```
Comment (mitchelloharawild) [inline modules/data-wrangling/index.qmd:308]: Maybe some suggestion that ignoring NAs should be done with care - something about the statistics and potential importance of understanding why the NAs are there?

This also seems more relevant to the earlier R basics sections, and a distraction from the core **data** manipulation operations being shown.

File: modules/data-wrangling/index.qmd [hunk 13]
```diff
@@ -165,11 +168,301 @@ For the expression field, there are many functions and operators that are useful
...
+:::
+
+:::{.callout-tip}
+## Did you know
+
+You may have noticed that we use the letter **z** in words like **summarize**. In British English, you might be more familiar with spelling it as **summarise**. In tidyverse, you can use either spelling when calling a function. So, whether you're used to writing summarize or summarising, the functionality is still the same either way.
```
Comment (mitchelloharawild) [inline modules/data-wrangling/index.qmd:314]: A bit wordy and possibly unimportant. Users will accidentally type these things the way they're familiar with and it won't matter.

**Review by mitchelloharawild — COMMENTED** [review]
Some comments on the work done so far. These suggestions also apply to other parts of the work - not just where I've attached the comment.

I think it'd be useful to think about the suggestions now, but revisit and fix them after the first 4 sections in R basics is done.
